_posts/2020-11-09-bash-pipes.md

```diff
@@ -0,0 +1,59 @@
+---
+layout: post
+title:  "Bash pipes: more useful than Mario's"
+date:   2020-11-09 12:54:04 -0700
+categories: bash pipes
+---
+You’re a budding bioinformatician and you are learning all of these amazing commands in bash to explore and modify your data. But you want a way to combine these commands without creating an endless amount of output files to work upon. Or you are trying to modify your data in a particular way, but you can't find that one *perfect* program that would accomplish your task.
+
+Luckily for you, Unix shell has modular workflows, allowing for programs to talk to one another, through **pipes**.
+
+Not unlike the pipes in Super Mario, these pipes can easily transport your data from one command to another. However, unlike Mario's pipes, bash pipes are simple to master and won't send you unexpectedly to your villain's lair.
+
+![mario pipe](https://thumbs.gfycat.com/FreeThoughtfulAmericancrayfish-poster.jpg)
+
+### How it works   
+Bash pipes are denoted with the `|` symbol, not to be confused with a lower case L or capital i.
+
+Instead of displaying the output on the terminal screen or redirecting the output to a new file (with `>`), a bash pipe will feed the output of one command directly to the input of another. As long as your commands can understand the same type of data, then you can use this on almost any Unix program. Bash pipes allow you to combine simple commands to make them into more complex tools to analyze your data.
+
+```Bash
+$ program1 inputfile.txt | program2
+```
+
+### Pros of modular workflows
+* Computational Efficiency
+  * reading and writing to the disk can be very slow particularly when using large datasets such as sequence data
+* Helpful in Debugging
+  * because each component is independent, it is easier to find mistakes in intermediate results
+* Customizable
+  * easily swap and add components
+* Optimizable
+  * combine the best-suited programs for your specific set of tasks
+
+### Cons
+* A pipe is only as good as it's weakest link
+  * If one part of your pipe-line has a bug, the entire thing will fail
+  * But, because it is modular, it is easy to identify the component that is bugged by testing each part in succession (see above in Pros)
+
+
+### Tee Pipes
+It may be useful to see intermediate outputs in a file format to identify bugs in the pipeline. With the `tee` program, you can divert a copy of your pipeline's output in a file while still passing it to the next step, just like a plumbers T-pipe.
+
+```Bash
+$ program1 inputfile.txt | tee intermediate-file.txt | program2 > results.txt
+```
+
+### Common examples
```

Comment:
I like all the examples you gave in this section. Maybe it could be a good idea to give a practical example to help the reader visualize how the pipe would work using the examples you gave. For example:
"$ head -n 20 inputfile.txt | cut -f 1,4
-> this would hand the first 20 lines of the file to the next program, and then 'cut -f 1,4' would retrieve the 1st and 4th columns of those 20 lines".